File: sdk/typescript/src/namespaces/plugins.ts
/**
 * Plugins Namespace API
 *
 * Provides a namespaced interface for plugin management.
 * Essential for extensibility and marketplace integrations.
 */

/**
 * Plugin definition
 */
export interface Plugin {
  id: string;
  name: string;
  description: string;
  version: string;
  author: string;
  category: string;
  tags: string[];
  icon_url?: string;
  documentation_url?: string;
  installed: boolean;
  enabled: boolean;
  config?: Record<string, unknown>;
  permissions: string[];
  created_at: string;
  updated_at: string;
}

/**
 * Plugin marketplace listing
 */
export interface PluginListing {
  id: string;
  name: string;
  description: string;
  version: string;
  author: string;
  category: string;
  tags: string[];
  icon_url?: string;
  rating: number;
  install_count: number;
  verified: boolean;
  featured: boolean;
}

/**
 * Plugin submission for marketplace review
 */
export interface PluginSubmission {
  id: string;
  plugin_name: string;
  version: string;
  status: 'pending' | 'reviewing' | 'approved' | 'rejected';
  submitted_at: string;
  reviewed_at?: string;
  reviewer_notes?: string;
}

/**
 * Plugin configuration schema
 */
export interface PluginConfigSchema {
  properties: Record<string, {
    type: string;
    description: string;
    required?: boolean;
    default?: unknown;
  }>;
}

/**
 * Install plugin request
 */
export interface InstallPluginRequest {
  config?: Record<string, unknown>;
}

/**
 * Plugin query request
 */
export interface PluginQueryRequest {
  query: string;
  category?: string;
  tags?: string[];
  limit?: number;
  offset?: number;
}

/**
 * Plugin validation request
 */
export interface PluginValidateRequest {
  manifest: Record<string, unknown>;
  code?: string;
}

/**
 * Interface for the internal client used by PluginsAPI.
 */
interface PluginsClientInterface {
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
  put<T>(path: string, body?: unknown): Promise<T>;
  delete<T>(path: string): Promise<T>;
}

/**
 * Plugins API namespace.
 *
 * Provides methods for managing plugins:
 * - Browse and search marketplace
 * - Install and configure plugins
 * - Run plugin functionality
 * - Submit plugins for review
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Browse marketplace
 * const { plugins, featured } = await client.plugins.getMarketplace();
 *
 * // Search plugins
 * const results = await client.plugins.query({ query: 'slack', category: 'integrations' });
 *
 * // Install a plugin
 * const plugin = await client.plugins.install('slack-connector', { config: { workspace: 'my-team' } });
 *
 * // Run plugin functionality
 * const result = await client.plugins.run('slack-connector', { action: 'send', channel: '#general' });
 * ```
 */
export class PluginsAPI {
  constructor(private client: PluginsClientInterface) {}

  // ===========================================================================
  // Plugin Discovery
  // ===========================================================================

  /**
   * List all available plugins.
   */
  async list(): Promise<{ plugins: Plugin[]; total: number }> {
    return this.client.get('/api/plugins');
  }

  /**
   * Get a specific plugin by name.
   */
  async get(name: string): Promise<Plugin> {
    return this.client.get(`/api/plugins/${name}`);
  }

  /**
   * Get marketplace listings with categories and featured plugins.
   */
  async getMarketplace(): Promise<{ plugins: PluginListing[]; categories: string[]; featured: PluginListing[] }> {
    return this.client.get('/api/plugins/marketplace');
  }

  /**
   * Query plugins with search and filters.
   */
  async query(body: PluginQueryRequest): Promise<{ plugins: PluginListing[]; total: number }> {
    return this.client.post('/api/plugins/query', body);
  }

  // ===========================================================================
  // Plugin Management
  // ===========================================================================

  /**
   * List installed plugins.
   */
  async listInstalled(): Promise<{ plugins: Plugin[]; total: number }> {
    return this.client.get('/api/plugins/installed');
  }

  /**
   * Install a plugin.
   */
  async install(name: string, body?: InstallPluginRequest): Promise<Plugin> {
    return this.client.post(`/api/plugins/${name}/install`, body);
  }

  /**
   * Uninstall a plugin.
   */
  async uninstall(name: string): Promise<{ uninstalled: boolean }> {
    return this.client.delete(`/api/plugins/${name}`);
  }

  /**
   * Run a plugin with the given parameters.
   */
  async run(name: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.post(`/api/plugins/${name}/run`, body);
  }

  // ===========================================================================
  // Plugin Submission
  // ===========================================================================

  /**
   * Submit a plugin for marketplace review.
   */
  async submit(body: Record<string, unknown>): Promise<PluginSubmission> {
    return this.client.post('/api/plugins/submit', body);
  }

  /**
   * List your plugin submissions.
   */
  async listSubmissions(): Promise<{ submissions: PluginSubmission[] }> {
    return this.client.get('/api/plugins/submissions');
  }

  /**
   * Validate a plugin manifest before submission.
   */
  async validate(body: PluginValidateRequest): Promise<{ valid: boolean; errors?: string[] }> {
    return this.client.post('/api/plugins/validate', body);
  }
}

File: sdk/python/aragora_sdk/namespaces/plugins.py
"""
Plugins Namespace API

Provides methods for plugin management:
- List and search plugins
- Install and run plugins
- Marketplace operations
- Plugin submissions
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class PluginsAPI:
    """
    Synchronous Plugins API.

    Provides methods for managing plugins that extend
    Aragora's functionality.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> plugins = client.plugins.list()
        >>> client.plugins.install("my-plugin")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Plugin Listing
    # ===========================================================================

    def list(self, category: str | None = None) -> dict[str, Any]:
        """
        List available plugins.

        Args:
            category: Filter by category

        Returns:
            Dict with plugins array
        """
        params: dict[str, Any] = {}
        if category:
            params["category"] = category

        return self._client.request("GET", "/api/v1/plugins", params=params)

    def list_installed(self) -> dict[str, Any]:
        """
        List installed plugins.

        Returns:
            Dict with installed plugins array
        """
        return self._client.request("GET", "/api/v1/plugins/installed")

    def get(self, name: str) -> dict[str, Any]:
        """
        Get details for a specific plugin.

        Args:
            name: Plugin name

        Returns:
            Plugin details
        """
        return self._client.request("GET", f"/api/v1/plugins/{name}")

    # ===========================================================================
    # Plugin Operations
    # ===========================================================================

    def install(self, name: str, version: str | None = None) -> dict[str, Any]:
        """
        Install a plugin.

        Args:
            name: Plugin name
            version: Version to install (latest if not specified)

        Returns:
            Installation result
        """
        data: dict[str, Any] = {}
        if version:
            data["version"] = version

        return self._client.request("POST", f"/api/v1/plugins/{name}/install", json=data)

    def uninstall(self, name: str) -> dict[str, Any]:
        """
        Uninstall a plugin.

        Args:
            name: Plugin name

        Returns:
            Uninstallation result
        """
        return self._client.request("DELETE", f"/api/v1/plugins/{name}/install")

    def run(
        self,
        name: str,
        action: str | None = None,
        params: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Run a plugin action.

        Args:
            name: Plugin name
            action: Action to run (default action if not specified)
            params: Parameters for the action

        Returns:
            Action result
        """
        data: dict[str, Any] = {}
        if action:
            data["action"] = action
        if params:
            data["params"] = params

        return self._client.request("POST", f"/api/v1/plugins/{name}/run", json=data)

    # ===========================================================================
    # Marketplace
    # ===========================================================================

    def list_marketplace(
        self,
        category: str | None = None,
        sort: str = "popular",
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List plugins in the marketplace.

        Args:
            category: Filter by category
            sort: Sort order (popular, newest, updated)
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Dict with plugins array and total count
        """
        params: dict[str, Any] = {"sort": sort, "limit": limit, "offset": offset}
        if category:
            params["category"] = category

        return self._client.request("GET", "/api/v1/plugins/marketplace", params=params)

    # ===========================================================================
    # Plugin Submissions
    # ===========================================================================

    def submit(
        self,
        name: str,
        description: str,
        repository_url: str,
        category: str,
        version: str = "1.0.0",
    ) -> dict[str, Any]:
        """
        Submit a plugin to the marketplace.

        Args:
            name: Plugin name
            description: Plugin description
            repository_url: Git repository URL
            category: Plugin category
            version: Initial version

        Returns:
            Submission result with submission_id
        """
        data: dict[str, Any] = {
            "name": name,
            "description": description,
            "repository_url": repository_url,
            "category": category,
            "version": version,
        }

        return self._client.request("POST", "/api/v1/plugins/submit", json=data)

    def list_submissions(self, status: str | None = None) -> dict[str, Any]:
        """
        List your plugin submissions.

        Args:
            status: Filter by status (pending, approved, rejected)

        Returns:
            Dict with submissions array
        """
        params: dict[str, Any] = {}
        if status:
            params["status"] = status

        return self._client.request("GET", "/api/v1/plugins/submissions", params=params)


class AsyncPluginsAPI:
    """
    Asynchronous Plugins API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     plugins = await client.plugins.list()
        ...     await client.plugins.install("my-plugin")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # Plugin Listing
    async def list(self, category: str | None = None) -> dict[str, Any]:
        """List available plugins."""
        params: dict[str, Any] = {}
        if category:
            params["category"] = category

        return await self._client.request("GET", "/api/v1/plugins", params=params)

    async def list_installed(self) -> dict[str, Any]:
        """List installed plugins."""
        return await self._client.request("GET", "/api/v1/plugins/installed")

    async def get(self, name: str) -> dict[str, Any]:
        """Get details for a specific plugin."""
        return await self._client.request("GET", f"/api/v1/plugins/{name}")

    # Plugin Operations
    async def install(self, name: str, version: str | None = None) -> dict[str, Any]:
        """Install a plugin."""
        data: dict[str, Any] = {}
        if version:
            data["version"] = version

        return await self._client.request("POST", f"/api/v1/plugins/{name}/install", json=data)

    async def uninstall(self, name: str) -> dict[str, Any]:
        """Uninstall a plugin."""
        return await self._client.request("DELETE", f"/api/v1/plugins/{name}/install")

    async def run(
        self,
        name: str,
        action: str | None = None,
        params: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Run a plugin action."""
        data: dict[str, Any] = {}
        if action:
            data["action"] = action
        if params:
            data["params"] = params

        return await self._client.request("POST", f"/api/v1/plugins/{name}/run", json=data)

    # Marketplace
    async def list_marketplace(
        self,
        category: str | None = None,
        sort: str = "popular",
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List plugins in the marketplace."""
        params: dict[str, Any] = {"sort": sort, "limit": limit, "offset": offset}
        if category:
            params["category"] = category

        return await self._client.request("GET", "/api/v1/plugins/marketplace", params=params)

    # Plugin Submissions
    async def submit(
        self,
        name: str,
        description: str,
        repository_url: str,
        category: str,
        version: str = "1.0.0",
    ) -> dict[str, Any]:
        """Submit a plugin to the marketplace."""
        data: dict[str, Any] = {
            "name": name,
            "description": description,
            "repository_url": repository_url,
            "category": category,
            "version": version,
        }

        return await self._client.request("POST", "/api/v1/plugins/submit", json=data)

    async def list_submissions(self, status: str | None = None) -> dict[str, Any]:
        """List your plugin submissions."""
        params: dict[str, Any] = {}
        if status:
            params["status"] = status

        return await self._client.request("GET", "/api/v1/plugins/submissions", params=params)
